backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/model/validator/M_ProductPrice.java

```diff
@@ -21,51 +35,103 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
- * 
+ *
  * You should have received a copy of the GNU General Public
  * License along with this program.  If not, see
  * <http://www.gnu.org/licenses/gpl-2.0.html>.
  * #L%
  */
 
 /**
- * Model validator <b>with attached callout</b>. See {@link #unsetIsHUPrice(I_M_ProductPrice)} for what it does.
- * 
+ * Model validator <b>with attached callout</b>. See {@link #validateUOM(I_M_ProductPrice)} for what it does.
+ *
  * @author ts
  * @task http://dewiki908/mediawiki/index.php/09045_Changes_on_Excel_Export-Import_%28107395757529%29
- *
  */
 @Callout(I_M_ProductPrice.class)
 @Validator(I_M_ProductPrice.class)
 public class M_ProductPrice
 {
+	private static final AdMessageKey MSG_ERR_M_PRODUCT_PRICE_MISSING_PACKING_ITEM = AdMessageKey.of("MSG_ERR_M_PRODUCT_PRICE_MISSING_PACKING_ITEM");
+	private static final AdMessageKey MSG_ERR_M_PRODUCT_PRICE_PACKING_ITEM_INFINITE_CAPACITY = AdMessageKey.of("MSG_ERR_M_PRODUCT_PRICE_PACKING_ITEM_INFINITE_CAPACITY");
+	private static final AdMessageKey MSG_ERR_M_PRODUCT_PRICE_NO_UOM_CONVERSION = AdMessageKey.of("MSG_ERR_M_PRODUCT_PRICE_NO_UOM_CONVERSION");
+
+	private final IUOMDAO uomDAO = Services.get(IUOMDAO.class);
+	private final IHUPIItemProductBL huPIItemProductBL = Services.get(IHUPIItemProductBL.class);
+	private final IHUCapacityBL huCapacityBL = Services.get(IHUCapacityBL.class);
+	private final IUOMConversionDAO uomConversionDAO = Services.get(IUOMConversionDAO.class);
+	private final IProductBL productBL = Services.get(IProductBL.class);
+
 	@Init
 	public void registerCallout()
 	{
 		Services.get(IProgramaticCalloutProvider.class).registerAnnotatedCallout(this);
 	}
-	
-	/**
-	 * We now allow PIIPs with unlimited capacity, but in that case, we make sure that the <code>M_ProductPrice_Attribute</code> is not a <code>HUPrice</code>.
-	 * @param productPrice
-	 */
-	@ModelChange(timings = ModelValidator.TYPE_BEFORE_CHANGE,
-			ifColumnsChanged = I_M_ProductPrice.COLUMNNAME_M_HU_PI_Item_Product_ID)
-	@CalloutMethod(columnNames = { I_M_ProductPrice.COLUMNNAME_M_HU_PI_Item_Product_ID })
-	public void unsetIsHUPrice(final I_M_ProductPrice productPrice)
+
+	@ModelChange(timings = { ModelValidator.TYPE_BEFORE_NEW, ModelValidator.TYPE_BEFORE_CHANGE },
+			ifColumnsChanged = I_M_ProductPrice.COLUMNNAME_C_UOM_ID)
+	@CalloutMethod(columnNames = { I_M_ProductPrice.COLUMNNAME_C_UOM_ID })
+	public void validateUOM(final I_M_ProductPrice productPrice)
 	{
-		if (productPrice.getM_HU_PI_Item_Product_ID() <= 0)
+		final UomId uomId = UomId.ofRepoId(productPrice.getC_UOM_ID());
+
+		if (uomDAO.isUOMForTUs(uomId))
 		{
-			productPrice.setIsHUPrice(false);
+			validateTuUom(productPrice);
+
+			productPrice.setIsHUPrice(true);
 		}
-		else if (productPrice.getM_HU_PI_Item_Product().isInfiniteCapacity())
+		else
 		{
 			productPrice.setIsHUPrice(false);
 		}
+
+		validatePriceUOM(productPrice);
```

Comment:
i think if it's *not* a TU-UOM, then we shouldn't execute this check..right?

Reply:
changed this. please check if it's OK

Reply:
I'll test the change and come back with an answer